src/compiler/binder.ts

```diff
@@ -352,6 +374,272 @@ namespace ts {
+
+            currentReachabilityState = ifTrueState;
+
+            // bind expression (don't affect reachability)
+            bind(n.expression);
+
+            bind(n.thenStatement);
+            if (n.elseStatement) {
+                const preElseState = currentReachabilityState;
+                currentReachabilityState = ifFalseState;
+                bind(n.elseStatement);
+                currentReachabilityState = or(currentReachabilityState, preElseState);
+            }
+            else {
+                currentReachabilityState = or(currentReachabilityState, ifFalseState);
+            }
+
+            return true;
+        }
+
+        function bindReturnOrThrow(n: ReturnStatement | ThrowStatement): boolean {
+            // bind expression (don't affect reachability)
+            bind(n.expression);
+            if (n.kind === SyntaxKind.ReturnStatement) {
+                hasExplicitReturn = true;
+            }
+            currentReachabilityState = Reachability.Unreachable;
+
+            return true;
+        }
+
+        function bindBreakOrContinueStatement(n: BreakOrContinueStatement): boolean {
+            // call bind on label (don't affect reachability)
+            bind(n.label);
+            // for continue case touch label so it will be marked a used
+            const isValidJump = jumpToLabel(n.label, n.kind === SyntaxKind.BreakStatement ? currentReachabilityState : Reachability.Unreachable);
+            if (isValidJump) {
+                currentReachabilityState = Reachability.Unreachable;
+            }
+            return true;
+        }
+
+        function bindTryStatement(n: TryStatement): boolean {
+            // catch\finally blocks has the same reachability as try block
+            const preTryState = currentReachabilityState;
+            bind(n.tryBlock);
+            const postTryState = currentReachabilityState;
+
+            currentReachabilityState = preTryState;
+            bind(n.catchClause);
+            const postCatchState = currentReachabilityState;
+
+            currentReachabilityState = preTryState;
+            bind(n.finallyBlock);
+
+            // post catch/finally state is reachable if
+            // - post try state is reachable - control flow can fall out of try block
+            // - post catch state is reachable - control flow can fall out of catch block
+            currentReachabilityState = or(postTryState, postCatchState);
+
+            return true;
+        }
+
+        function bindSwitchStatement(n: SwitchStatement): boolean {
+            const preSwitchState = currentReachabilityState;
+            const postSwitchLabel = pushImplicitLabel();
+
+            // bind expression (don't affect reachability)
+            bind(n.expression);
+
+            bind(n.caseBlock);
+
+            const hasDefault = forEach(n.caseBlock.clauses, c => c.kind === SyntaxKind.DefaultClause);

```

Comment:
It's fine for parent nodes to set state based on their children.  That's because if a cihld is even touched by an edit, the parent will *always* be recreated.  So, conceptually there would be no incremental problem with adding such a flag to the 'switch' node.  That said, we've also tried to keep AST nodes pretty lightweight and consistent.  We don't want everyone hanging data off them willy nilly.  Also, even if we store the data on the node, we're just pushing the perf cost elsewhere.  After all, the parser will then have to do the same determination if there's a default clause as it doesn't really "know" during parsing.  After all, the clauses are just parsed during the recursive child parsing loop, and aren't done specially inside the switch parsing code.